xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -1360,4 +1364,37 @@ static boolean matchHostName(String hostName, String pattern) {
     return index == pattern.length() - 1
         && hostName.startsWith(pattern.substring(0, pattern.length() - 1));
   }
+
+  /**
+   * Convert protobuf message to human readable String format. Useful for protobuf messages
+   * containing {@link com.google.protobuf.Any} fields.
+   */
+  @VisibleForTesting
+  static class MessagePrinter {
+    private final JsonFormat.Printer printer;
+
+    @VisibleForTesting
+    MessagePrinter() {
+      com.google.protobuf.TypeRegistry registry =
+          com.google.protobuf.TypeRegistry.newBuilder()
+              .add(Listener.getDescriptor())
+              .add(HttpConnectionManager.getDescriptor())
+              .add(RouteConfiguration.getDescriptor())
+              .add(Cluster.getDescriptor())
+              .add(ClusterLoadAssignment.getDescriptor())
+              .build();
+      printer = JsonFormat.printer().usingTypeRegistry(registry);
+    }
+
+    @VisibleForTesting
+    String print(MessageOrBuilder message) {
+      String res;
+      try {
+        res = printer.print(message);
+      } catch (InvalidProtocolBufferException e) {
+        res = "Failed to convert message [" + message + "] to readable String. Reason: " + e;
```

Comment:
This language will be a bit confusing in the logs. Maybe something closer to: `message + " (failed to pretty-print: " + e + ")"`

Reply:
Changed. Thanks for the suggestion.

Reply:
Why not just return `message.toString()`? And the error will happen again and can be logged in `sendNackRequest()` in error detail `"Broken LDS response.  " + e`